storage/src/main/java/org/commonjava/storage/pathmapped/core/PathDBOutputStream.java

```diff
@@ -74,16 +89,19 @@
     }
 
     @Override
-    public void write( int b ) throws IOException
+    public void write (byte[] b, int off, int len)
+    throws IOException
     {
+        if ((off | len | (b.length - (len + off)) | (off + len)) < 0)
+            throw new IndexOutOfBoundsException();
+
         try
         {
-            super.write( b );
-            size += 1;
-            byte by = (byte) ( b & 0xff );
+            out.write( b, off, len );
+            size += len;
             if ( checksumCalculator != null )
             {
-                checksumCalculator.update( by );
+                checksumCalculator.update( b );
```

Comment:
should this be a subarray of b? cal by off, len.

Reply:
MessageDigest update method accepts an array. Which uses the length of the array and a 0 offset. Avoiding repeated calls to MessageDigest.update for every individual byte.

Reply:
That is the reason I concerned. Some of the bytes in the array might have already been digested. That is why out.write only write a sub-array. If I don't miss anything, the digest should use the same sub-array.

Reply:
This PR has not changed the [existing api](https://github.com/Commonjava/path-mapped-storage/blob/master/storage/src/main/java/org/commonjava/storage/pathmapped/util/ChecksumCalculator.java#L35-L43) of ChecksumCalculator. The code modification has swapped the method being used.
 I think any user who calls [update(byte[])](https://github.com/Commonjava/path-mapped-storage/blob/master/storage/src/main/java/org/commonjava/storage/pathmapped/util/ChecksumCalculator.java#L35) is expecting every element of the array to be used. By my understanding it is the responsibility of the api caller to create a new sub-array and pass it to ChecksumCalculator to update using every element of the array.
 Are you proposing the ChecksumCalculator should expose a new update(byte[], int, int) method to address your concerns ? I've not found any places in the code-base that require that functionality yet and would appreciate understanding better the root cause of this concern.

Reply:
It is about the write(byte[], off, len). 
The prev code did not override it because the filteredoutputstream.write(byte[], off, len) call write(b) so we don't worry about it. The new code extends the outputstream where the write(byte[], off, len) came in and use its own way. We have to override it (as you 've done). This method's byte[] array is like a buffer and the real bytes we should use is the sub-array identified by [off,len]. That is why "out.write( b, off, len )". For same reason, we should call sth like "checksumCalculator.update( b, off, len )". If you simply pass the whole array to digester, the digester result will be wrong. The easy way is to add sth as below. 
`+++ b/storage/src/main/java/org/commonjava/storage/pathmapped/util/ChecksumCalculator.java
+    public final void update( final byte[] data, final int offset, final int len )
+    {
+        digester.update( data, offset, len );
+    }
`

Reply:
Yes I see what you mean.